File: packages/rn/ios/DashOtaBundleLoader.swift
import Foundation

/// The early bundle-resolution hook. The host app's Swift `AppDelegate` calls this from
/// `bundleURL()` (release branch only) so the active OTA bundle (or embedded fallback) is
/// chosen before React starts. Runs the crash-loop circuit breaker. Returns nil to fall back
/// to the embedded bundle (fail closed).
///
/// Usage in the host `AppDelegate.swift`:
/// ```
/// import DashOta
/// override func bundleURL() -> URL? {
///   #if DEBUG
///   return RCTBundleURLProvider.sharedSettings().jsBundleURL(forBundleRoot: "index")
///   #else
///   return DashOtaBundleLoader.bundleURL() ?? Bundle.main.url(forResource: "main", withExtension: "jsbundle")
///   #endif
/// }
/// ```
@objc(DashOtaBundleLoader)
public class DashOtaBundleLoader: NSObject {
  /// The active OTA bundle URL, or nil to fall back to the embedded bundle.
  @objc public static func bundleURL() -> URL? {
    guard let path = DashOtaStore.shared.resolveBundleAtLaunch() else { return nil }
    return URL(fileURLWithPath: path)
  }
}

File: packages/rn/android/src/main/java/com/dashota/DashOtaBundleLoader.kt
package com.dashota

import android.content.Context

/**
 * The early bundle-resolution hook. The host app calls this from
 * `ReactNativeHost.getJSBundleFile()` so the active OTA bundle (or embedded fallback) is
 * chosen **before** React starts. Runs the crash-loop circuit breaker. Never throws — on any
 * error it returns null so the app falls back to the embedded bundle (fail closed).
 *
 * Usage in the host `MainApplication.kt`:
 * ```
 * override fun getJSBundleFile(): String? =
 *   DashOtaBundleLoader.getBundleFile(applicationContext)
 * ```
 */
object DashOtaBundleLoader {
  @JvmStatic
  fun getBundleFile(context: Context): String? = try {
    DashOtaStore.resolveBundleAtLaunch(context)
  } catch (_: Exception) {
    null
  }
}
